scipy/stats/_unuran/unuran_wrapper.pyx.templ

```diff
@@ -0,0 +1,1075 @@
+    provide an algorithm that is not too slow and needs only a short setup.
+    This method can be applied to continuous unimodal distributions with
+    T-concave density function. See [1]_ and [2]_ for more details.
+
+    Parameters
+    ----------
+    dist : object
+        An instance of a class with 'pdf' and 'dpdf' methods.
+
+        * 'pdf': PDF of the distribution. The signature of the PDF is
+          expected to be: ``def pdf(x: float, *params) -> float``. i.e.
+          the PDF should accept a Python float and the parameters and
+          return a Python float. It doesn't need to integrate to 1 i.e.
+          the PDF doesn't need to be normalized.
+        * 'dpdf': Derivative of the PDF w.r.t x (i.e. the variate). Must
+          have the same signature as the PDF.
+
+    mode : float, optional
+        Mode of the distribution. It is optional and only required if
+        `use_mode` is ``True``.
+    center : float, optional
+        Center (``loc``) of the distribution. It is optional and only
+        required if `use_center` is ``True``. Default is 0.
+    params : tuple, optional
+        Parameters that the PDF and DPDF take. Default is an empty tuple.
+    domain : list or tuple of length 2, optional
+        The support of the distribution. Must not contain `nan` values.
+        Default is ``None``. When ``None``, the support is assumed to be
+        :math:`(-\infty, \infty)`.
+    c : float, optional
+        The transformation function. See Notes for the allowed values for
+        this parameter and their interpretation. Default is -0.5.
+    cpoints : int, optional
+        The number of construction points. Default is 30.
+    variant : str, optional
+        The variant to use. Available options are:
+
+          * 'ps': PS variant (Default)
+          * 'gw': GW variant
+          * 'ia': IA variant
+
+        See Notes for their explainations.
+    use_dars : bool, optional
+        If True, "derandomized adaptive rejection sampling" (DARS) is used
+        in setup. See [1]_ for the details of the DARS algorithm. Default
+        is True.
+    max_sqhratio : float, optional
+        Set upper bound for the ratio (area below squeeze) / (area below hat).
+        It must be a number between 0 and 1. Default is 0.99.
+    max_intervals : int, optional
+        Set maximum number of intervals. Default is 100.
+    use_center : bool, optional
+        Use the center as construction point, if given. Default is True.
+    use_mode : bool, optional
+        Use the (exact!) mode as construction point, if given. Default is
+        True.
+    guide_factor : float, optional
+        Set factor for relative size of the guide table for indexed search.
+        It must be greater than or equal to 0. When set to 0, then sequential
+        search is used. Default is 2.
+    seed : int, BitGenerator, Generator, RandomState, SeedSequence, optional
+        Seed for the underlying uniform random number generation.
```

Comment:
I propose we use the same in QMC. We should not show it works with `RandomState`. If the use pass a `RandomState` it should use it, otherwise we should not create one and only use a `Generator`.

```suggestion
    seed : {None, int, `numpy.random.Generator`}, optional
        If `seed` is None the `numpy.random.Generator` singleton is used.
        If `seed` is an int, a new ``Generator`` instance is used,
        seeded with `seed`.
        If `seed` is already a ``Generator`` instance then that instance is
        used.
```

Reply:
> I propose we use the same in QMC. We should not show it works with `RandomState`. If the use pass a `RandomState` it should use it, otherwise we should not create one and only use a `Generator`.

OK. Sounds good to me. Do you agree, @chrisb83 @mckib2?

Reply:
This change seems in line with gh-13778, so LGTM

Reply:
I disagree. Describe accurately and completely what the argument accepts.

Reply:
> I disagree. Describe accurately and completely what the argument accepts.

True, but as I suggested in another comment, I would not even allow `RandomState` as we are going to move to NumPy > 1.16.

Reply:
Change the code first, then document that code, not vice versa.

Also, there is no `Generator` singleton.

Reply:
> Also, there is no `Generator` singleton.

This is what we wrote in QMC. I will change it then. Should it just read ```If `seed` is None the `numpy.random.Generator` is used.```?

Reply:
1. Consider changing the name of the argument to `rng` instead of `seed` (@tupui also in the QMC code). Generally, we don't want people to pass in seed integers to these APIs. Except in the most trivial of programs (mostly examples), these are not going to be the only APIs that will be using the PRNG. So we want people to build their `Generator` once and pass that in.

2. Even when we do drop support for 1.16, I think these APIs should still accept `RandomState` instances (with documentation to match). We'll just reach inside to grab its `BitGenerator` and wrap a `Generator` around it to actually use instead of passing it along untouched.

Reply:
> Should it just read `` If `seed` is None the `numpy.random.Generator` is used. ``?

`default_rng()` describes the case thusly: `If None, then fresh, unpredictable entropy will be pulled from the OS.`

Reply:
> 1. Consider changing the name of the argument to `rng` instead of `seed` (@tupui also in the QMC code). Generally, we don't want people to pass in seed integers to these APIs. Except in the most trivial of programs (mostly examples), these are not going to be the only APIs that will be using the PRNG. So we want people to build their `Generator` once and pass that in.

This makes sense to do. For this particular case, an even better name would be `urng` as in uniform random number generated. What do you think @tupui @rkern?

Reply:
I think it will be better to use the same name everywhere in scipy where the semantics are the same (like we mostly standardized on `random_state` for `RandomState`). In any case, while you're only using the uniform subset of the API, we're still asking for a `Generator`.

Reply:
> I think it will be better to use the same name everywhere in scipy where the semantics are the same (like we mostly standardized on `random_state` for `RandomState`). In any case, while you're only using the uniform subset of the API, we're still asking for a `Generator`.

OK, that sounds good. I will change it to `rng` then. Thanks!

Reply:
Wait a bit. It will be best to coordinate with the QMC stuff, and that may require a deprecation round, so we'll have to discuss a little. I don't want you to have to undo and redo a bunch of stuff.

Reply:
> Wait a bit. It will be best to coordinate with the QMC stuff, and that may require a deprecation round, so we'll have to discuss a little. I don't want you to have to undo and redo a bunch of stuff.

Sure! I will try to keep myself updated with the discussions on QMC.

Reply:
> 1. Consider changing the name of the argument to `rng` instead of `seed` (@tupui also in the QMC code). Generally, we don't want people to pass in seed integers to these APIs. Except in the most trivial of programs (mostly examples), these are not going to be the only APIs that will be using the PRNG. So we want people to build their `Generator` once and pass that in.

It's a shame we did not do that before 1.7... Now we would need a cycle to deprecate it. We added this in a few other places as well. As far as I can see, we have more occurrence of `seed` than `random_state` and just one of `rng`.

> 2. Even when we do drop support for 1.16, I think these APIs should still accept `RandomState` instances (with documentation to match). We'll just reach inside to grab its `BitGenerator` and wrap a `Generator` around it to actually use instead of passing it along untouched.

This should be done in NumPy by `np.random.default_rng` directly no? If yes, I would be happy to propose a PR.

I propose to continue this discussion here https://github.com/scipy/scipy/issues/14322